### PR TITLE
[Feature] Support Phi Loss

### PR DESCRIPTION
**Pull request by androbaza**
## Motivation

Support PhiLoss as described in https://doi.org/10.3390/s21082803. The implementation is published by authors of the paper in this [repo](https://github.com/rubeea/focal_phi_loss_mmsegmentation), but the pull request was never created for that. I fixed some parts of the implementation which were outdated for the current version of mmseg and cleaned up commented lines used during development. 

This Loss is particularly useful in case of highly unbalanced datasets.

## Modification

Add PhiLoss class, add the class and function in respective `__init__.py` file. 

## BC-breaking 

None

## Use cases (Optional)

Tested this loss during my own dataset training, helped me to improve the imbalanced dataset issue without manually introducing class weights.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMDet3D.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by MengzhangLI** [comment]
Hi, @androbaza thanks for your nice PR. We would review your PR as soon as possible!

BTW, could you attach your experiments on this new loss? Like what I did in [Tversky loss PR](https://github.com/open-mmlab/mmsegmentation/pull/1986#issue-1353638378)?

**Comment by androbaza** [comment]
> Hi, @androbaza thanks for your nice PR. We would review your PR as soon as possible!
> 
> BTW, could you attach your experiments on this new loss? Like what I did in [Tversky loss PR](https://github.com/open-mmlab/mmsegmentation/pull/1986#issue-1353638378)?

Hi @MengzhangLI, I will definitely add some results as soon as I have finalized the experiments (may take several weeks). I have a custom dataset, though, but I will describe the situation and why I came to using Phi Loss.

**Comment by MengzhangLI** [comment]
> m

OK. Thank you so much. Could you also try to use Phi Loss on our provided dataset such as DRIVE, STARE, HRF and CHASE_DB1? They are all medical vessel segmentation datasets.

**Comment by MengzhangLI** [comment]
> > Hi, @androbaza thanks for your nice PR. We would review your PR as soon as possible!
> > BTW, could you attach your experiments on this new loss? Like what I did in [Tversky loss PR](https://github.com/open-mmlab/mmsegmentation/pull/1986#issue-1353638378)?
> 
> Hi @MengzhangLI, I will definitely add some results as soon as I have finalized the experiments (may take several weeks). I have a custom dataset, though, but I will describe the situation and why I came to using Phi Loss.

OK. Thank you so much. Could you also try to use Phi Loss on our provided dataset such as DRIVE, STARE, HRF and CHASE_DB1? They are all medical vessel segmentation datasets.

File: mmseg/models/losses/phi_loss.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    loss= reduce_loss(loss,reduction=reduction)
+    return loss
+
+
+@LOSSES.register_module()
+class PhiLoss(nn.Module):
+    """PhiLoss.
+
+    Args:
+        reduction (str, optional): . Defaults to 'mean'.
```
Comment (MeowZheng) [inline mmseg/models/losses/phi_loss.py:36]: It would be better to complete the description of `reduction`

File: mmseg/models/losses/phi_loss.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        self.smooth = smooth
+        self._loss_name = loss_name
+
+    def forward(self,
+                inputs,
+                targets,
+                weight=None,
+                reduction_override=None,
+                ignore_index=255):
+        """Forward function."""
```
Comment (MeowZheng) [inline mmseg/models/losses/phi_loss.py:61]: I suggest add the docstring for these arguments to explain their function

Comment (androbaza) [reply mmseg/models/losses/phi_loss.py:66]: Hi, the function is indeed identical to the original that you have at BCE:

https://github.com/open-mmlab/mmsegmentation/blob/ee25adc2c1d73e22355d168fa09ae7bdac899a24/mmseg/models/losses/cross_entropy_loss.py#L254

The function could be understood by those who want to do it from checking the ` _expand_onehot_labels` function in the original BCE:

https://github.com/open-mmlab/mmsegmentation/blob/ee25adc2c1d73e22355d168fa09ae7bdac899a24/mmseg/models/losses/cross_entropy_loss.py#L66

Honestly, I would love to provide a description for what's happening, but I'm not able to as I am not the one who developed these parts.

Comment (MeowZheng) [reply mmseg/models/losses/phi_loss.py:66]: ok I'll add the docstring for all `forward` function after this pr merged

**Review by MeowZheng — COMMENTED** [review]
I suggest using the pre-commit hook to sovle the lint problem, you might refer [this doc](https://github.com/open-mmlab/mmcv/blob/master/docs/en/community/pr.md#pr-specs),
and would you like to add unit test for Phi loss? Here is an example for unit test of ce loss https://github.com/open-mmlab/mmsegmentation/blob/master/tests/test_models/test_losses/test_ce_loss.py 

Thanks for your contribution for mmseg!

**Comment by androbaza** [comment]
> I suggest using the pre-commit hook to sovle the lint problem, you might refer [this doc](https://github.com/open-mmlab/mmcv/blob/master/docs/en/community/pr.md#pr-specs), and would you like to add unit test for Phi loss? Here is an example for unit test of ce loss https://github.com/open-mmlab/mmsegmentation/blob/master/tests/test_models/test_losses/test_ce_loss.py
> 
> Thanks for your contribution for mmseg!

Hi, I got the lint problem solved. 
Unfortunately I have a super tight schedule so far and can not create the unit test in foreseeable future -- but i will definitely remember about this the next time i have some time to contribute. Hope it helps.

**Comment by MeowZheng** [comment]
> > I suggest using the pre-commit hook to sovle the lint problem, you might refer [this doc](https://github.com/open-mmlab/mmcv/blob/master/docs/en/community/pr.md#pr-specs), and would you like to add unit test for Phi loss? Here is an example for unit test of ce loss https://github.com/open-mmlab/mmsegmentation/blob/master/tests/test_models/test_losses/test_ce_loss.py
> > Thanks for your contribution for mmseg!
> 
> Hi, I got the lint problem solved. Unfortunately I have a super tight schedule so far and can not create the unit test in foreseeable future -- but i will definitely remember about this the next time i have some time to contribute. Hope it helps.

I have added `Phi Loss` in our `2.0.0rc1` milestone which will be released on Dec 1st. Does this time work for you?

**Comment by androbaza** [comment]
> > > I suggest using the pre-commit hook to sovle the lint problem, you might refer [this doc](https://github.com/open-mmlab/mmcv/blob/master/docs/en/community/pr.md#pr-specs), and would you like to add unit test for Phi loss? Here is an example for unit test of ce loss https://github.com/open-mmlab/mmsegmentation/blob/master/tests/test_models/test_losses/test_ce_loss.py
> > > Thanks for your contribution for mmseg!
> > 
> > 
> > Hi, I got the lint problem solved. Unfortunately I have a super tight schedule so far and can not create the unit test in foreseeable future -- but i will definitely remember about this the next time i have some time to contribute. Hope it helps.
> 
> I have added `Phi Loss` in our `2.0.0rc2` milestone which will be released on Dec 1st. Does this time work for you?

Sure, thank you!
